### DISCUSSION TITLE
Splitting a image into an array of tiles and the reverse?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16817] [Q&A]
I am having trouble figuring out how to split a image into an array of smaller tiles.
I am also wondering how to opposite can be acheaved.

**Comment by stevemonaco** [comment]
This isn't an Avalonia task. Avalonia's role, IMO, is to load and render images. Specific image manipulations, not including visual effect filters/shaders, is outside of scope.

You should pick a library that is capable of loading, creating, and manipulating images, such as ImageSharp. Then write separate logic so that Avalonia can render it.

**Reply by Shadowblitz16** [reply]
How would I then pass that image to avalonia for loading though?

**Reply by stevemonaco** [reply]
You don't. You either: 1. write code to render it with `SKBitmap` with SkiaSharp or 2. write code to copy it into a `WriteableBitmap` which then gets assigned to an `Image.Source` (or invalidated for an update). If you aren't using a compatible pixel format, then you'll also need to transcode the pixel format.

**Reply by Shadowblitz16** [reply]
> You don't. You either: 1. write code to render it with `SKBitmap` with SkiaSharp or 2. write code to copy it into a `WriteableBitmap` which then gets assigned to an `Image.Source` (or invalidated for an update). If you aren't using a compatible pixel format, then you'll also need to transcode the pixel format.

Gross. 
Why doesn't avalonia allow passing rgba32 pixel data directly to an image?

**Reply by stevemonaco** [reply]
It's like 20 LoC to do it from scratch. If you can't write that, you probably won't be successful in writing a full app. Few apps need to directly render pixel buffers. You may be able to get away with creating a `WriteableBitmap` and using `CopyPixels`.

**Reply by Shadowblitz16** [reply]
@stevemonaco you have to use unsafe code to use WriteableBitmap.

**Reply by stevemonaco** [reply]
That's only true if you're transcoding a pixel buffer and want efficiency. Otherwise you should be able to use `CopyPixels` or `Marshal.Copy`, neither of which require unsafe code. Writing basic unsafe code isn't something to be avoided when there's a clear benefit.

**Reply by dbriard** [reply]
You can do something like that to convert a SKBitmap to Bitmap: 
```
public static Bitmap ToAvalonia(this SKBitmap bitmap)
{
    if (!TryGetPixelFormat(bitmap.ColorType, bitmap.AlphaType, out var pf, out var af))
    {
        throw new NotSupportedException();
    }

    var info = bitmap.Info;
    return new Bitmap(pf, af, bitmap.GetAddress(0, 0), new PixelSize(info.Width, info.Height), new Vector(96, 96), info.RowBytes);
}
```

The big problem is that Avalonia Bitmap do not have something like SKColorSpace, and so Colors could be incorrects. May be for V12? (I really hope).

If your images are using various color profiles, you need more code to convert them to Avalonia for display.
(or you can implement a custom control with ICustomDrawOperation to display SKBitmap or SKImage, instead if using Avalonia Image control with Bitmap).
